[refs] Return `:ident`s in QP `annotate`; save in `:result_metadata`

bshepherdson opened the pull request:
[refs] Return `:ident`s from annotate mw; save in `result_metadata`

Fixes [QUE-701](https://linear.app/metabase/issue/QUE-701), [QUE-778](https://linear.app/metabase/issue/QUE-778),  [QUE-777](https://linear.app/metabase/issue/QUE-777),  [QUE-776](https://linear.app/metabase/issue/QUE-776),  [QUE-703](https://linear.app/metabase/issue/QUE-703),  [QUE-704](https://linear.app/metabase/issue/QUE-704).

### Description

Directly implements `:ident` calculation for legacy MBQL in the QP!

This is an unfortunately large PR, but I don't think much can be done to help that.
The output of `annotate` becomes `Card.result_metadata`, which is a key **input** to
`annotate`. So we have to tackle legacy metadata end to end to produce a healthy PR.

(A prior attempt to use the lib and match up the results turned into a cluster of hacks,
and required half of these changes anyway, so this direct approach seems more sound.)

Most of this logic lives in the `annotate` post-processing middleware, with
some support from `add-source-metadata` pre-processing.


#### Idents

This generates correct idents for proper fields, expressions,
aggregations, breakouts, explicit and implicit joins, native queries and
models (native and MBQL).

#### Safety checks and Guarantees

Also adds and strengthens several safety checks in the library and
elsewhere. The guarantees I hope we can count on at this point are:
- Cards are saved with correct `:ident`s in their `:result_metadata`.
- Cards with old `:result_metadata` get (correct) `:ident`s added on read.
- Lib can now rely on getting correct idents for all input cards!
- Therefore the lib (BE + FE) can compute the ident for **every column**.

#### Placeholder idents

For native queries, we can only get `result_metadata` from executing them.
The usual flow is:
- Write the query
- Send it as an ad-hoc `/api/dataset` request
- Get back `result_metadata`
- Save it, passing the `result_metadata` back to the BE.

We need a card `entity_id` to create idents for a native query, but an ad-hoc
query doesn't have a card yet! The BE now generates a placeholder `entity_id`,
which gets replaced on save with the new card's actual `entity_id`. These
placeholders contain (1) illegal NanoID characters, and (2) a NanoID, which
means they cannot be mistaken for non-placeholder NanoID, and that they are
still globally unique, so there's no caching concerns.

### How to verify

Tests, naturally. There's sufficient safety checks in both BE and lib to object
if some idents are missing or malformed, when they're expected everywhere.

- You can see the `:ident`s in the `data.results_metadata.columns` and
  `data.cols` on query responses.
- Also in the `result_metadata` of a card response.
- Run an ad-hoc native query to see the placeholders in its idents
    - And save it to see that they get replaced correctly.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

bshepherdson committed:
[refs] Return `:ident`s from annotate mw; save in `result_metadata`

Directly implements `:ident`s handling for legacy MBQL between the
`add-source-metadata` preprocessing middleware and the `annotate`
post-processing middleware.

This generates correct idents for proper fields, expressions,
aggregations, breakouts, explicit and implicit joins, native queries and
models (native and MBQL).

Also adds and strengthens several safety checks in the library and
elsewhere. The guarantees I hope we can count on at this point are:
- Cards are saved with `:ident`s in their `:result_metadata`.
- Cards with old `:result_metadata` get `:ident`s added on read.
- Lib can now rely on getting correct idents for all input cards.
- Therefore Lib can compute the ident for **every column**.

bshepherdson committed:
model.card fixes to require idents in result_metadata

Failing some tests that reveal at least some gaps in `annotate`, so
backtracking to that now.

bshepherdson committed:
Fixing models.card-test; needs all tests to be run

bshepherdson committed:
Placeholder idents! End-to-end consistent idents for native queries

bshepherdson committed:
poking around to fix model idents getting double-bagged

bshepherdson committed:
Cleaning up debugging noise in annotate etc.

bshepherdson committed:
linter cleanup

bshepherdson committed:
tidying some debugging output etc.

bshepherdson commented:
* **#56244** <a href="https://app.graphite.dev/github/pr/metabase/metabase/56244?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/metabase/metabase/56244?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

bshepherdson committed:
Test fixes - api/dataset_test.clj is in progress next

metamben commented on src/metabase/query_processor/middleware/annotate.clj:
```diff
@@ -271,8 +275,17 @@
-                  :display_name (humanization/name->human-readable-name id-or-name)}))
+                 ;; HACK: If the column can be found in the join's :source-metadata, use its :ident but still fall
+                 ;; back to the naive case.
+                 #_(u/prog1 (some-> (some #(when (= (:name %) id-or-name) %) (:source-metadata join))
+                                    (dissoc :field_ref))
+                     (tap> (list `col-info-for-field-clause* "string ref found in join" clause '=> <>)))
```
This should be resolved.

bshepherdson replied:
Yeah, it's sorted out now.

metamben commented on src/metabase/query_processor/middleware/annotate.clj:
```diff
@@ -298,6 +311,20 @@
+      (update :ident lib/explicitly-joined-ident (:ident join))
+
+      ;; Sometimes we see an implicitly joined ref with `:source-field`, but no corresponding join clause.
+      ;; In that case, look up the FK's ident and use adjust this column's ident.
```
```suggestion
      ;; In that case, look up the FK's ident and use it to adjust this column's ident.
```

metamben commented on src/metabase/query_processor/middleware/annotate.clj:
```diff
@@ -455,7 +484,21 @@
+     (select-keys source-metadata-col [:unit]))
+
+   ;; Idents are sometimes preserved from the source, and sometimes replaced.
+   ;; TODO: Expressions the simply reference fields directly keep their IDs and might also need special handling here?
```
Can't parse this sentence.

bshepherdson replied:
How's that?

metamben commented on src/metabase/query_processor/middleware/annotate.clj:
```diff
@@ -637,7 +715,7 @@
   "Middleware for adding type information about the columns in the query results (the `:cols` key)."
   [{query-type :type, :as query
-    {:keys [:metadata/model-metadata :alias/escaped->original]} :info} rff]
+    {:keys [card-entity-id :metadata/model-metadata :alias/escaped->original]} :info} rff]
```
I find it easier to recognize names if they look like symbols (especially with color coding):
```suggestion
    {:keys [card-entity-id metadata/model-metadata alias/escaped->original]} :info} rff]
```

metamben commented on src/metabase/query_processor/middleware/constraints.clj:
```diff
@@ -66,6 +67,11 @@
   (update query :constraints (comp ensure-valid-constraints merge-default-constraints)))
 
+(defn- add-card-entity-id
+  "Userland queries are always associated with a `:card-entity-id`. If one was not provided, generate a placeholder."
```
```suggestion
  "Userland queries are always associated with a `:card-entity-id`, so we generate and add a placeholder."
```

metamben commented on test/metabase/channel/render/card_test.clj:
```diff
@@ -206,31 +207,41 @@
                                   model-query    :dataset_query
                                   model-metadata :result_metadata
                                   :as            model-card} {:type            :model
+                                                              :entity_id       "zQ1h5VDeOchxwMDYSK7om"
```
I would prefer naming `"zQ1h5VDeOchxwMDYSK7om"`, e.g., as `entity-id`. Generating a random one would be the clearest, IMO.

metamben commented on test/metabase/lib/metadata/calculation_test.cljc:
```diff
@@ -266,77 +266,58 @@
-        (is (= []
-               (->> (lib/visible-columns query)
-                    (remove (comp #{:source/card} :lib/source)))))))
+  (letfn [(query-with-user-id-tweaks [tweaks]
+            (let [base     (-> (lib.tu/mock-cards)
+                               :orders/native
+                               lib.tu/as-model)
+                  metadata (mapv (fn [col]
+                                   (cond-> col
+                                     (= (:name col) "USER_ID") (merge tweaks)))
+                                 (:result-metadata base))
+                  model    (assoc base :result-metadata metadata)]
```
Nice!

metamben commented on test/metabase/query_processor/util/add_alias_info_test.clj:
```diff
@@ -856,23 +859,25 @@
-                                                [:field "B1" {:base-type :type/Integer, :join-alias "Model B - A1"}]],
-                                               :source-table "card__2"}],
-                                             :source-table "card__1"}}}]}))))
+               (let [eid (u/generate-nano-id)]
```
Is this supposed to be reused somewhere?

bshepherdson replied:
I think I was just following the shape of the ones above, but yeah, it's not reused.

bshepherdson committed:
Test fixes and debugging. Checkpoint before tweaking models in annotate!

bshepherdson committed:
More test fixes, not quite done

appleby commented on src/metabase/lib/metadata/ident.cljc:
```diff
@@ -40,6 +43,92 @@
 
+;; ## Placeholder :idents
+;; We need the `:entity_id` of the card for native and model idents, but ad-hoc queries don't have an `:entity_id`
+;; yet. In that case we generate a **placeholder** `:entity_id` which gets replaced when the card is saved.
```
Placeholders are only needed for models based on native queries, I take it? When I first read this comment, I was thinking it would apply to either native or mbql models, but when I try it out locally, I seem to only get placeholder idents for ad-hoc native model's columns.

bshepherdson replied:
The comment is accurate. When you send an ad-hoc native query (question or model) it's necessary. They're also used for an ad-hoc model (MBQL or native) before you've saved it for the first time.

appleby replied:
Hmm well now I'm not getting placeholder idents even for ad-hoc native queries. It seems like something ([Quesiton.ts](https://github.com/metabase/metabase/blob/1c4abe5ae99d52141bf8db5a441b8637e4c5a0bb/frontend/src/metabase-lib/v1/Question.ts#L936)?) on the FE is passing along a `:card-entity-id` with the request, and that is being respected. Maybe this is expected, but to clarify what I'm doing:

* New > SQL query
* Enter anything, say `SELECT * FROM ORDERS`
* Click the "play" button (Run query)
* Generates a `/dataset` request where the request payload contains `info: {"card-entity-id": "lJ0CWHZNWODa83HYnWCLw"}`
* Response contains `context: "ad-hoc"` and e.g. `data.cols[0].ident = "native__lJ0CWHZNWODa83HYnWCLw__ID"` (no ADHOC placeholder)
* Save the question
* couple of requests are made when saving, but e.g. the response from the `/card` endpoint
    * `entity_id: "Jma-9JyNf0cggwpWmRPEM"`
    * `dataset_query.info["card-entity-id"]: "lJ0CWHZNWODa83HYnWCLw"`
    * `result_metadata[0].ident: "native__lJ0CWHZNWODa83HYnWCLw__ID"`

Now create new native question

* New > SQL query
* Enter something different, say `SELECT ID FROM PEOPLE`
* Click the "play" button (Run query)
* Generates a `/dataset` request where the request payload contains the same `card-entity-id` as before `info: {"card-entity-id": "lJ0CWHZNWODa83HYnWCLw"}`
* Response contains `context: "ad-hoc"` and e.g. `data.cols[0].ident = "native__lJ0CWHZNWODa83HYnWCLw__ID"` (same as previous ad-hoc query, but this is `PEOPLE.ID` vs `ORDERS.ID` from the prev query)
* Save the question
* response from the `/card` endpoint
    * `entity_id: "Z0d0zs1lwWEDh0dY6158N"`
    * `dataset_query.info["card-entity-id"]: "lJ0CWHZNWODa83HYnWCLw"`
    * `result_metadata[0].ident: "native__lJ0CWHZNWODa83HYnWCLw__ID"`


Occasionally (after restart?), I do get the placeholder idents like `"native__$$ADHOC[a2iOJS17N7RFKlicJqOPM]__ID"`, but I was expecting them every time for (e.g.) ad-hoc native queries. When those placeholders are generated, they do get replaced with the card `entity_id` on save.

bshepherdson replied:
Those outputs are very concerning; it's definitely wrong for the idents on a native query to have a different `entity_id` in them than the card has.

I do wonder if either (1) your had a URL hash with that reused `"lJ0CWHZNWODa83HYnWCLw"` in it from the old version, or (2) the CLJS and JS didn't get fully rebuilt properly.

I'll try to reproduce this.

appleby replied:
Yeah, it's also weird that the ident seems to be "sticky" between different questions. I did try various things like closing/re-opening the browser in case there was browser state getting cached, killing and restarting my yarn build, building a jar and running that in case something was weird about my dev build, blowing away my app db, etc. Let me know if you can't reproduce, maybe my build is just in a bad state.

appleby replied:
Pulled latest changes and confirm I'm consistently seeing placeholders for ad-hoc native queries now.

appleby commented on src/metabase/lib/metadata/ident.cljc:
```diff
@@ -40,6 +43,92 @@
+(defn replace-placeholder-idents
+  "Given an `:ident` and the true `:entity_id` for a card, overwrite the placeholders."
+  [ident card-entity-id]
+  (str/replace ident placeholder-regex card-entity-id))
```
I got NPE here due to `ident` being null when I launched this in my repl starting from an old v50 database I had lying around. Haven't dug into it yet, but posting here in case it's expected or the cause is obvious to you.

Last few log lines from repl

```
2025-04-11 04:13:19,987 INFO audit-app.audit :: Adjusted Audit DB for loading Analytics Content
2025-04-11 04:13:19,988 INFO audit-app.audit :: The app is not running from a jar, starting copy...
2025-04-11 04:13:19,988 INFO audit-app.audit :: Copying /Users/ma/src/repos/metabase/refs-overhaul-idents-in-annotate-direct/resources/instance_analytics -> /Users/ma/src/repos/metabase/refs-overhaul-idents-in-annotate-direct/plugins/instance_analytics
2025-04-11 04:13:20,041 INFO audit-app.audit :: Copying complete.
2025-04-11 04:13:20,048 INFO audit-app.audit :: Loading Analytics Content from: /Users/ma/src/repos/metabase/refs-overhaul-idents-in-annotate-direct/plugins/instance_analytics
Execution error (NullPointerException) at metabase.lib.metadata.ident/replace-placeholder-idents (ident.cljc:75).
Cannot invoke "Object.toString()" because "s" is null
```


top of the stack trace

```
1. Caused by java.lang.NullPointerException
   Cannot invoke "Object.toString()" because "s" is null

                string.clj:  101  clojure.string/replace
                string.clj:   75  clojure.string/replace
                ident.cljc:   75  metabase.lib.metadata.ident$replace_placeholder_idents/invokeStatic
                ident.cljc:   72  metabase.lib.metadata.ident$replace_placeholder_idents/invoke
                  core.clj: 6261  clojure.core/update
                  core.clj: 6251  clojure.core/update
              metadata.clj:  232  metabase.models.card.metadata/fix-incoming-idents/fn
                  core.clj: 2759  clojure.core/map/fn/fn
           ArrayChunk.java:   60  clojure.lang.ArrayChunk/reduce
             protocols.clj:  135  clojure.core.protocols/fn
             protocols.clj:  123  clojure.core.protocols/fn
             protocols.clj:   19  clojure.core.protocols/fn/G
             protocols.clj:   31  clojure.core.protocols/seq-reduce
             protocols.clj:   74  clojure.core.protocols/fn
             protocols.clj:   74  clojure.core.protocols/fn
             protocols.clj:   13  clojure.core.protocols/fn/G
                  core.clj: 7026  clojure.core/transduce
                  core.clj: 7042  clojure.core/into
                  core.clj: 7029  clojure.core/into
              metadata.clj:  232  metabase.models.card.metadata/fix-incoming-idents
              metadata.clj:  226  metabase.models.card.metadata/fix-incoming-idents
                  core.clj: 6261  clojure.core/update
                  core.clj: 6251  clojure.core/update
              metadata.clj:  251  metabase.models.card.metadata/populate-result-metadata
              metadata.clj:  236  metabase.models.card.metadata/populate-result-metadata
```

bshepherdson replied:
Yeah, I ran into that case too. There's a `when` now.

bshepherdson committed:
Updating serdes test data and fixing a few other tests

bshepherdson committed:
A few more test fixes! Now only things that are known broken locally?

bshepherdson committed:
lint cleanup

bshepherdson committed:
cleaning up lint leftover commentary/debugging

bshepherdson commented on src/metabase/models/card.clj:
```diff
@@ -795,6 +800,84 @@
+(defn- backfill-result-metadata-idents*
+  [result-metadata {query :dataset_query, eid :entity_id, :as card}]
+  (case (:type query)
+    ;; For native queries, we only need to
```
Fixed.

bshepherdson commented on src/metabase/query_processor/middleware/annotate.clj:
```diff
@@ -477,32 +520,56 @@
+
+  cols
+  ;; XXX: Leaving this as a breadcrumb during development.
+  #_(for [col cols]
```
It's back in the latest version.

bshepherdson commented on test/metabase/query_processor/middleware/annotate_test.clj:
```diff
@@ -226,20 +266,24 @@
                     :display-name "Child"
                     :parent-id    2)]}))
 
+;; XXX: START HERE: Continue adding idents to these tests.
```
Ha, yeah, I did continue filling in this file.

bshepherdson committed:
Addressing code review comments

wzimrin reviewed (approved):
Looks reasonable to me, though I'm definitely not familiar enough with some of these files to fully understand the changes.

bshepherdson committed:
Lots of test fixes, I think everything except the display_name issue

bshepherdson committed:
BE unit tests passing aside from the new broken one

bshepherdson committed:
Dropping the broken test

bshepherdson committed:
Remove the hacky `card-entity-id` generation on the FE

It's no longer necessary; the placeholders on the BE are getting the
work done.

bshepherdson committed:
formatting and lint cleanups

bshepherdson committed:
adjusting the spelling of the idents to `native[card-eid]__...` etc.

appleby commented on src/metabase/query_processor/middleware/annotate.clj:
```diff
@@ -482,32 +526,69 @@
+  [cols card-entity-id]
+  ;; Safety checks during dev and test.
+  (when-not config/is-prod?
+    (when-not (every? :ident cols)
```
`(every lib/valid-basic-ident? cols)` ?
